File: TP02/Filme.java
import java.util.*;
import java.io.*;
import java.text.SimpleDateFormat;

public class Filme{

    public static final String FILE_PATH = "/tmp/disneyplus.csv";
    public static ArrayList<Filme> todosFilmes = new ArrayList<Filme>();

    private String Id;
    private String Tipo;
    private String Titulo;
    private String Diretor;
    private ArrayList<String> Cast;
    private String Pais;
    private Date Data;
    private int Ano;
    private String Rating;
    private String Duracao;
    private ArrayList<String> Listado;

    // Contrutor vazio
    public Filme(){

        this.Id = "";
        this.Tipo = "";
        this.Titulo = "";
        this.Diretor = "";
        this.Cast = new ArrayList<String>();
        this.Pais = "";
        this.Data = null;
        this.Ano = 0;
        this.Rating = "";
        this.Duracao = "";
        this.Listado = new ArrayList<String>();
    }

    // Construtor
    public Filme(String id, String tipo, String titulo, String diretor, 
    ArrayList<String> cast, String pais, Date data, int ano, String rating, String duracao, ArrayList<String> listado){

        Id = id;
        Tipo = tipo;
        Titulo = titulo;
        Diretor = diretor;
        Cast = cast;
        Pais = pais;
        Data = data;
        Ano = ano;
        Rating = rating;
        Duracao = duracao;
        Listado = listado;
    }

    // Gets
    public String getId() { return this.Id; }
    public String getTipo() { return this.Tipo; }
    public String getTitulo() { return this.Titulo; }
    public String getDiretor() { return this.Diretor; }
    public String getPais() { return this.Pais; }
    public Date getData() { return this.Data; }
    public int getAno() { return this.Ano; }
    public String getRating() { return this.Rating; }
    public String getDuracao() { return this.Duracao; }
    public String getListado() {

        String list = "[";

        for(int i = 0; i < this.Listado.size(); i++){
            list += this.Listado.get(i);
            if(i < this.Listado.size() - 1)
                list += ", ";
        }

        list += "]";

        if(list == "[]")
            list = "[NaN]";

        return list;
    }
    public String getCast(){
        
        String cast = "[";

        for(int i = 0; i < this.Cast.size(); i++) {

            cast += this.Cast.get(i);
            if(i < this.Cast.size() - 1) 
                cast += ", ";
        }

        cast += "]";

        if(cast == "[]")
            cast = "[NaN]";

        return cast;
    }

    // Sets
    public void setId(String id) { this.Id = id; }
    public void setTipo(String tipo) { this.Tipo = tipo; }
    public void setTitulo(String titulo) { this.Titulo = titulo; }
    public void setDiretor(String diretor) { this.Diretor = diretor; }
    public void setPais(String pais) { this.Pais = pais; }
    public void setData(Date data) { this.Data = data; }
    public void setAno(int ano) { this.Ano = ano; }
    public void setRating(String rating) { this.Rating = rating; }
    public void setDuracao(String duracao) { this.Duracao = duracao; }
    public void setListado(ArrayList<String> listado) { this.Listado = listado; }
    public void setCast(ArrayList<String> cast) { this.Cast = cast; }

    // Clone
    public Filme Clone(){
        return new Filme(this.Id, this.Tipo, this.Titulo, this.Diretor, this.Cast, 
        this.Pais, this.Data, this.Ano, this.Rating, this.Duracao, this.Listado);
    }

    public void print() {
   
        //se a data não for nula, formata no padrão "MMMM d, yyyy" do arquivo csv
        //caso contrário, exibe "NaN" para indicar que não havia data
        String dataAdicionada = (Data != null) ?
        new SimpleDateFormat("MMMM d, yyyy", Locale.ENGLISH).format(Data) 
        : "March 1, 1900";

        //converte o ano de lançamento para string para exibição
        String anoLancamento = String.valueOf(Ano);
    
        //impressão das imformações fomatadas
        System.out.println(
            "=> " + Id +
            " ## " + (Titulo.equals("NaN") ? "NaN" : Titulo) +
            " ## " + (Tipo.equals("NaN") ? "NaN" : Tipo) +
            " ## " + (Diretor.equals("NaN") ? "NaN" : Diretor) +
            " ## " + getCast() +
            " ## " + (Pais.equals("NaN") ? "NaN" : Pais) +
            " ## " + dataAdicionada +
            " ## " + anoLancamento +
            " ## " + (Rating.equals("NaN") ? "NaN" : Rating) +
            " ## " + (Duracao.equals("NaN") ? "NaN" : Duracao) +
            " ## " + getListado() + " ##"
        );
    }

    public void read(String linhaCsv) {
        String[] lista = new String[11];
        int i = 0, j = 0;
    
        while(j < 11) {  //enquanto não tiver os 11 campos preenchidos
            String campo = "";  //variável para armazenar o valor de cada campo
            
            //se o campo estiver entre aspas, processa o campo de maneira especial
            if(i < linhaCsv.length() && linhaCsv.charAt(i) == '"') {
                i++;  //pula a primeira aspa dupla
                
                //enquanto não encontrar a aspa final ou uma sequência de duas aspas duplas (representando uma aspa literal)
                while(i < linhaCsv.length()) {
                    if(linhaCsv.charAt(i) == '"') {  //se encontrar uma aspa dupla
                        //verifica se são duas aspas duplas seguidas, que devem ser tratadas como uma aspa literal
                        if(i + 1 < linhaCsv.length() && linhaCsv.charAt(i + 1) == '"') {
                            i += 2;  //pula as duas aspas
                        }
                        else {  //se for a aspa final, sai do loop
                            i++;  //pula a aspa final
                            break;
                        }
                    }
                    else {  //se não for uma aspa dupla, adiciona o caractere ao campo
                        campo += linhaCsv.charAt(i);
                        i++;
                    }
                }
                //pula a vírgula que separa os campos, se houver
                if(i < linhaCsv.length() && linhaCsv.charAt(i) == ',') i++;
            }
            else {
                //caso o campo não esteja entre aspas, vai até a vírgula ou o final da linha
                while(i < linhaCsv.length() && linhaCsv.charAt(i) != ',') {
                    campo += linhaCsv.charAt(i);
                    i++;
                }
                //pula a vírgula, se houver
                if(i < linhaCsv.length() && linhaCsv.charAt(i) == ',') i++;
            }
    
            //se o campo estiver vazio, considera como "NaN"
            if(campo.isEmpty()) campo = "NaN";
            
            //armazena o campo no array e incrementa o índice
            lista[j] = campo;
            j++;
        }
        
        //faz as atribuições dos valores
        this.Id = lista[0];
        this.Tipo = lista[1];
        this.Titulo = lista[2];
        this.Diretor = lista[3];
        //se for "NaN", salva como "NaN", se não, quebra a string sempre que encontrar vígula seguida de espaço
        if(lista[4].equals("NaN")){
            ArrayList<String> list = new ArrayList<>();
            list.add("NaN");
            this.Cast = list;
        }
        else{
            this.Cast = new ArrayList<>(Arrays.asList(lista[4].split(", ")));
        }
        this.Pais = lista[5];
        this.Data = safeParseDate(lista[6]); //converte em formato de data
        this.Ano = safeParseInt(lista[7]); //converte em inteiro
        this.Rating = lista[8];
        this.Duracao = lista[9];
        //se for "NaN", salva como "NaN", se não, quebra a string sempre que encontrar vígula seguida de espaço
        if(lista[10].equals("NaN")){
            ArrayList<String> list = new ArrayList<>();
            list.add("NaN");
            this.Listado = list;
        }
        else{
            this.Listado = new ArrayList<>(Arrays.asList(lista[10].split(", ")));
        }
    }

    private Date safeParseDate(String valor) {
        try {
            if(valor.equals("NaN")) return null; //se data for "NaN", retorna null

            //se data possuir valor, converte no formato do arquivo csv
            SimpleDateFormat formato = new SimpleDateFormat("MMMM d, yyyy", Locale.ENGLISH);
            return formato.parse(valor.trim());
        }
        catch(Exception e) {
            return null; //em caso de excessão, retorna null
        }
    }
    
    private int safeParseInt(String valor) {
        try {
            //converte o ano de lançamento para inteiro e usa trim() para remover espaços inesperados
            return Integer.parseInt(valor.trim());
        }
        catch(Exception e) {
            return 0; //em caso de excessão, retorna 0
        }
    }

    public static void LerFilmes() {

        try {

            FileInputStream fstream = new FileInputStream(FILE_PATH);
            BufferedReader br = new BufferedReader(new InputStreamReader(fstream));

            String linha = br.readLine();
  
            while((linha = br.readLine()) != null) {

                Filme filme = new Filme();
                filme.read(linha);
                todosFilmes.add(filme);
            }

            // Fechar CSV
            fstream.close();
        }
        catch(IOException e) { }
    }

    public static Filme getById(String id, ArrayList<Filme> filmes) {

        for(int i = 0; i < filmes.size(); i++) {

            if(filmes.get(i).getId().equals(id)) return filmes.get(i);
        }
        return null;
    }

    public static void main(String[] args) {

        //Ler todos os filmes em CSV
        LerFilmes();
        Scanner sc = new Scanner(System.in);
        Filme filme = new Filme();
        String linha = sc.nextLine();

        while(!linha.equals("FIM")) {

            // Get id
            String id = linha;

            // buscar filme
            filme = getById(id, todosFilmes);

            // Printar filme
            if(filme != null) 
                filme.print();

            linha = sc.nextLine();
        }

        sc.close();
    }
}

File: TP02/Filme.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_FILMES 1000
#define MAX_LINHA 1024

typedef struct {
    char id[100];
    char tipo[100];
    char titulo[100];
    char diretor[100];
    char cast[512];
    char pais[100];
    char data[100];
    int ano;
    char rating[50];
    char duracao[50];
    char listado[512];
} Filme;

Filme filmes[MAX_FILMES];
int totalFilmes = 0;

// Função para substituir strings vazias por "NaN"
void preencherNaN(char *campo) {
    if (strlen(campo) == 0 || strcmp(campo, "\n") == 0)
        strcpy(campo, "NaN");
}

// Função para remover aspas e limpar campo CSV
void limparCampo(char *campo) {
    char *src = campo, *dst = campo;
    while (*src) {
        if (*src != '"') {
            *dst++ = *src;
        }
        src++;
    }
    *dst = '\0';
}

void removerAspasDuplas(char *str) {
    char *src = str, *dst = str;
    while (*src) {
        if (*src == '"' && *(src + 1) == '"') {
            *dst++ = '"'; // uma aspa literal
            src += 2;
        } else if (*src == '"') {
            src++; // pula aspas normais
        } else {
            *dst++ = *src++;
        }
    }
    *dst = '\0';
}

// Função para ler e parsear o CSV
void parseLinha(char *linha, Filme *f) {
    int campo = 0;
    int i = 0, j = 0;
    char temp[1024] = "";
    int len = strlen(linha);
    int dentroAspas = 0;

    char campos[11][1024] = {0};

    while (i <= len) {
        char c = linha[i];
        if (c == '"' && linha[i + 1] == '"') {
            temp[j++] = '"';
            i += 2;
        }
        else if (c == '"') {
            dentroAspas = !dentroAspas;
            i++;
        }
        else if ((c == ',' || c == '\0' || c == '\n') && !dentroAspas) {
            temp[j] = '\0';
            strcpy(campos[campo++], temp);
            j = 0;
            temp[0] = '\0';
            i++;
        }
        else {
            temp[j++] = c;
            i++;
        }
    }

    removerAspasDuplas(campos[0]); strcpy(f->id, campos[0]);
    removerAspasDuplas(campos[1]); strcpy(f->tipo, campos[1]);
    removerAspasDuplas(campos[2]); strcpy(f->titulo, strlen(campos[2]) ? campos[2] : "NaN");
    removerAspasDuplas(campos[3]); strcpy(f->diretor, strlen(campos[3]) ? campos[3] : "NaN");
    removerAspasDuplas(campos[4]); strcpy(f->cast, strlen(campos[4]) ? campos[4] : "NaN");
    removerAspasDuplas(campos[5]); strcpy(f->pais, strlen(campos[5]) ? campos[5] : "NaN");
    removerAspasDuplas(campos[6]); strcpy(f->data, strlen(campos[6]) ? campos[6] : "NaN");
    removerAspasDuplas(campos[7]); f->ano = atoi(campos[7]);
    removerAspasDuplas(campos[8]); strcpy(f->rating, strlen(campos[8]) ? campos[8] : "NaN");
    removerAspasDuplas(campos[9]); strcpy(f->duracao, strlen(campos[9]) ? campos[9] : "NaN");
    removerAspasDuplas(campos[10]); strcpy(f->listado, strlen(campos[10]) ? campos[10] : "NaN");
}

void lerCSV(const char *caminho) {
    FILE *fp = fopen(caminho, "r");
    if (!fp) {
        return;
    }

    char linha[MAX_LINHA];
    fgets(linha, MAX_LINHA, fp); // Pula o cabeçalho

    while (fgets(linha, MAX_LINHA, fp)) {
        if (totalFilmes >= MAX_FILMES) break;
        Filme f;
        memset(&f, 0, sizeof(Filme));
        parseLinha(linha, &f);
        filmes[totalFilmes] = f;
        totalFilmes++;
    }

    fclose(fp);
}

// Função para buscar filme por ID
Filme* buscarFilmePorId(char* id) {
    for (int i = 0; i < totalFilmes; i++) {
        if (strcmp(id, filmes[i].id) == 0)
            return &filmes[i];
    }
    return NULL;
}

void ordenarStrings(char *itens[], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (strcmp(itens[j], itens[j + 1]) > 0) {
                char *temp = itens[j];
                itens[j] = itens[j + 1];
                itens[j + 1] = temp;
            }
        }
    }
}


void formatarListaOrdenada(const char *entrada, char *saida) {
    if (strcmp(entrada, "NaN") == 0) {
        strcpy(saida, "[NaN]");
        return;
    }

    // Copia da string original, porque strtok modifica
    char copia[512];
    strcpy(copia, entrada);

    // Array de ponteiros para os itens
    char *itens[100];
    int count = 0;

    char *token = strtok(copia, ",");
    while (token && count < 100) {
        // Remove espaços extras
        while (*token == ' ') token++;
        itens[count++] = token;
        token = strtok(NULL, ",");
    }

    // Ordena os itens
    ordenarStrings(itens, count);

    // Monta a string formatada
    strcpy(saida, "[");
    for (int i = 0; i < count; i++) {
        strcat(saida, itens[i]);
        if (i < count - 1) strcat(saida, ", ");
    }
    strcat(saida, "]");
}

// Função para imprimir filme formatado
void printFilme(Filme *f) {
    char castFormatado[600], listadoFormatado[600];

    formatarListaOrdenada(f->cast, castFormatado);
    formatarListaOrdenada(f->listado, listadoFormatado);

    printf("=> %s ## %s ## %s ## %s ## %s ## %s ## %s ## %d ## %s ## %s ## %s ##\n",
        f->id,
        f->titulo,
        f->tipo,
        f->diretor,
        castFormatado,
        f->pais,
        strlen(f->data) ? f->data : "March 1, 1900",
        f->ano,
        f->rating,
        f->duracao,
        listadoFormatado
    );
}

int main() {
    char input[100];
    lerCSV("/tmp/disneyplus.csv");

    // Primeira leitura
    fgets(input, sizeof(input), stdin);
    input[strcspn(input, "\n")] = '\0'; // Remove o \n

    while (strcmp(input, "FIM") != 0) {
        Filme *f = buscarFilmePorId(input);
        if (f)
            printFilme(f);

        // Lê a próxima entrada
        fgets(input, sizeof(input), stdin);
        input[strcspn(input, "\n")] = '\0';
    }

    return 0;
}